src/relay.cpp

```diff
@@ -0,0 +1,196 @@
+                                  output_topic(output_topic)
+  {
+    initialize();
+  }
+
+  void RelayNode::initialize()
+  {
+    lazy = declare_parameter<bool>("lazy", false);
+
+    auto ros_clock = rclcpp::Clock::make_shared();
+    timer_ = rclcpp::create_timer(this, ros_clock, LOOP_PERIOD, [=]()
+                                  { timer_callback(); });
+
+    // so that we execute the loop the first time instantly
+    timer_callback();
+  }
+
+  void RelayNode::timer_callback()
+  {
+    if (!topic_type)
+    {
+      if (topic_type = try_find_topic_type())
+      {
+        // we just found the topic type, so create the publisher and the subscriber
+        pub_ = this->create_generic_publisher(output_topic, *topic_type, rclcpp::QoS(1));
+        if (!lazy){
+          subscribe();
+        }
+      }
+    } 
+    else if (lazy)
+    {
+      const size_t sub_count = pub_->get_subscription_count();
+      if (sub_count > 0)
+      {
+        if (!sub_){
+          subscribe();
+        }
+      } else
+      {
+        if (sub_) {
+          unsubscribe();
+        }
+      }
+    }
+  }
+
+  /*
+Since we don't know the topic type in advance we need to determine it from
+either the input_topic or the output_topic. If only one of them exists, we
+use its type, but if both exist an arbitrary decision was made here to use
+the type of input_topic.
+*/
+  std::optional<std::string> RelayNode::try_find_topic_type()
+  {
+    auto const topic_names_and_types = this->get_topic_names_and_types();
+    auto topic_type_i = topic_names_and_types.find(input_topic);
+
+    if (topic_type_i != topic_names_and_types.end())
+    {
+      return topic_type_i->second[0];
+    }
+
+    topic_type_i = topic_names_and_types.find(output_topic);
+
+    if (topic_type_i != topic_names_and_types.end())
+    {
+      return topic_type_i->second[0];
+    }
+
+    return {};
+  }
+
+  void RelayNode::subscribe()
+  {
+    sub_ = this->create_generic_subscription(
+      input_topic, *topic_type, rclcpp::QoS(1), 
+      std::bind(&RelayNode::incoming_message_callback, this, std::placeholders::_1));
+  }
+
+  void RelayNode::unsubscribe()
+  {
+    if (sub_) {
+      sub_.reset();
+    }
+  }
+
+  void RelayNode::incoming_message_callback(std::shared_ptr<rclcpp::SerializedMessage> msg)
+  {
+    rclcpp::SerializedMessage msg_rcl = *msg.get();
+    pub_->publish(msg_rcl);
+  }
+
+} // namespace topic_tools
+
+int main(int argc, char *argv[])
+{
+  rclcpp::init(argc, argv);
+
+  std::shared_ptr<topic_tools::RelayNode> node;
+
+  if (argc >= 2 && strcmp(argv[1], "--ros-args"))
+  {
+    std::string output_topic;
+    if (argc >= 3 && strcmp(argv[2], "--ros-args"))
+    {
+      output_topic = argv[2];
+    }
+    else 
+    {
+      output_topic = std::string(argv[1]) + "_relay";
+    }
+    node = std::make_shared<topic_tools::RelayNode>("Relay", argv[1], output_topic);
+  }
+  else
+  {
+    node = std::make_shared<topic_tools::RelayNode>("Relay");
+  }
```

Comment:
I think probably don't bother doing commandline parsing - rclcpp::init will parse the ros args - and that already allows for node naming, parameters, etc. This ties back to what I was saying above about having a single configuration path.

```suggestion
  rclcpp::NodeOptions options;
  auto node = std::make_shared<topic_tools::RelayNode>(options);
```

Reply:
Following this suggestion would remove the ability to run `ros2 run topic_tools relay <intopic> [outtopic]`, so I think it's worth it to leave the parsing as it is, along with the ability to pass `input_topic` and `output_topic` as node parameters